### PR TITLE
Substantial rewrite of NovaCompute

### PR DESCRIPTION
**Pull request by mdbooth**
Details are in the various commit messages, but this amounts to a rewrite of NovaCompute. Let me know if there was any good reason not to use systemd for service control, btw. However, what was there didn't work, so I had to change it in order to be able to test it.

I am concerned that 'NovaCompute stop' only stops the compute service, and does not stop its running instances. This is not a change in behaviour, btw, as the code there previously to do migrations was disabled. It was also wrong, because list-migrations doesn't work that way.

I am not familiar with pacemaker, so my understanding of compute-managed.scenario is limited. However, I noticed that when fencing a nova compute we physically power it off before evacuating its instances, which is obviously a robust fencing method. 'NovaCompute stop' doesn't achieve the same result, because the compute service is down but the instances are still running. Hence evacuating them elsewhere is most definitely not safe. In face, the only safe thing to do if we aren't going to stonith is turn the compute service back on, which makes me doubt the utility of stopping it in the first place.

Let me know if you'd like a 'graceful complete shutdown of compute service and instances' behaviour for 'NovaCompute stop'. Also let me know if you'd like a migration option, which was hinted at in the usage text but not implemented.

It's worth noting, btw, that the current stonith behaviour is only safe when the hypervisor is co-located with the compute service. This is the case with the libvirt driver, but it is not the case for either the vmware or ironic drivers. You might want to consider documenting that this only works for co-located hypervisors. However, in those cases it is possible to stand up a backup nova compute which talks to the same hypervisor, i.e. evacuation is not necessary. In the case of the vmware driver you must ensure the nova computes don't run simultaneously, though.

### CODE REVIEW

**Comment by mdbooth** [comment]
fabbione pointed out that systemd's restart behaviour might be undesirable. I suggest that using systemd for process start/stop/status is still the most robust solution available, but agree that start/stop management must be in a single domain for a cluster, and that should be pacemaker.

I understand there is already a puppet module which disables the openstack-nova-compute service, presumably with 'systemctl disable openstack-nova-compute.service'. If that same puppet module drops a file into /etc/systemd/system/openstack-nova-compute.service.d/local.conf with the following contents:

  [Service]
  Restart=no

then systemd becomes a dumb, but very robust way to start and stop a pacemaker-managed openstack-nova-compute resource. NovaCompute will also no-longer have to duplicate the environment variable set in the systemd unit. Tested on RHOS 6/RHEL 7.

**Comment by mdbooth** [comment]
Alternatively, we could create a separate systemd resource to manage the systemd unit directly from pacemaker. I think the ordering of this works, because we're doing:

nova_stop:
- nova service stuff
- process stop

nova_start:
- process start
- nova service stuff

i.e. we've got a strict ordering with no interleaving => simple dependency.

Half-baked thought: is there any way we could just turn this into additional actions for systemd, and have pacemaker manage only the single systemd unit?

**Comment by fabbione** [comment]
On 05/21/2015 06:13 PM, mdbooth wrote:

> fabbione pointed out that systemd's restart behaviour might be
> undesirable. I suggest that using systemd for process start/stop/status
> is still the most robust solution available, but agree that start/stop
> management must be in a single domain for a cluster, and that should be
> pacemaker.
> 
> I understand there is already a puppet module which disables the
> openstack-nova-compute service, presumably with 'systemctl disable
> openstack-nova-compute.service'. If that same puppet module drops a file
> into /etc/systemd/system/openstack-nova-compute.service.d/local.conf
> with the following contents:
> 
> [Service]
> Restart=no

This already happens in background and it's done by pacemaker when
configuring any systemd: resource.

> then systemd becomes a dumb, but very robust way to start and stop a
> pacemaker-managed openstack-nova-compute resource. NovaCompute will also
> no-longer have to duplicate the environment variable set in the systemd
> unit. Tested on RHOS 6/RHEL 7.

I agree, we can just have NovaCompute resource-agent manage the migration:

ocf::openstack:NovaCompute then systemd:openstack-nova-compute

**Comment by beekhof** [comment]
> I am concerned that 'NovaCompute stop' only stops the compute service, and does not stop its running instances.

So am I, but there was a long thread and I was explicitly told not to implement it or at least not make that the default.  (The existing code should be considered pseudo code only, it has never been executed).

> Let me know if there was any good reason not to use systemd for service control, btw.

Essentially it comes down to compatibility. The set of agents to which this will belong is required to function across a large number of distros and distros versions - so we can't assume that systemd is present on the machine.

> It's worth noting, btw, that the current stonith behaviour is only safe when the hypervisor is co-located with the compute service. 

Interesting, I may have more questions on this.

> Half-baked thought: is there any way we could just turn this into additional actions for systemd, and have pacemaker manage only the single systemd unit?

This might have some merit, I'm going to think this over before I respond definitively :)

File: pcmk/NovaCompute [hunk 1]
```diff
@@ -108,150 +112,122 @@ END
...
+
+    # Wait for nova to find the service
+    nova_wait_for_service
+    rc=$?
+    if [ $rc != $OCF_SUCCESS ]; then
+        return $rc
+    fi
+
+    # Enable the service in nova
+    novaclient service_enable ${NOVA_HOST} nova-compute
```
Comment (vuntz) [inline pcmk/NovaCompute:190]: FWIW, I implemented it that way upstream: https://review.openstack.org/275929

**Comment by aspiers** [comment]
Can we close this or is it still relevant?

**Comment by mdbooth** [comment]
This can be closed.

Matt

**Comment by aspiers** [comment]
Pinging @beekhof to click the close button :-)

**Pull request closed without merging**